This notebook demonstrates how to use Apache Spark RDDs to process COVID-19
case and test data from New York State, including aggregation, positivity
rate calculations, and monthly trend analysis.

In [2]:
//COVID-19 Data Analysis with Apache Spark (New York State)

val counties = Array("Kings","Queens","New+York","Suffolk","Bronx","Nassau","Westchester","Erie",
                     "Monroe","Richmond","Onondaga","Orange","Rockland","Albany","Dutchess",
                     "Saratoga","Oneida","Niagara","Broome","Ulster","Rensselaer","Schenectady",
                     "Chautauqua","Oswego","Jefferson","Ontario","St.+Lawrence","Tompkins",
                     "Putnam","Steuben","Wayne","Chemung","Sullivan","Clinton","Cattaraugus",
                     "Cayuga","Madison","Warren","Columbia","Livingston","Washington","Herkimer",
                     "Otsego","Genesee","Fulton","Montgomery","Greene","Tioga","Franklin","Chenango",
                     "Cortland","Allegany","Delaware","Wyoming","Orleans"
                     ,"Essex","Seneca","Schoharie","Lewis","Yates","Schuyler","Hamilton")

val base_url = "https://health.data.ny.gov/resource/xdss-u53e.json?County="
val urls = counties.map(a => base_url+a) //Makes a url for each county


val results = urls.map(u => scala.io.Source.fromURL(u).mkString) 


val data_rdd = spark.read.json(sc.parallelize(results,32))
                    .rdd
                    .map(r => (r(5).toString.slice(0,10), 
                               r(0).toString,
                               r(4).toString.toInt,
                               r(7).toString.toInt))


counties = Array(Kings, Queens, New+York, Suffolk, Bronx, Nassau, Westchester, Erie, Monroe, Richmond, Onondaga, Orange, Rockland, Albany, Dutchess, Saratoga, Oneida, Niagara, Broome, Ulster, Rensselaer, Schenectady, Chautauqua, Oswego, Jefferson, Ontario, St.+Lawrence, Tompkins, Putnam, Steuben, Wayne, Chemung, Sullivan, Clinton, Cattaraugus, Cayuga, Madison, Warren, Columbia, Livingston, Washington, Herkimer, Otsego, Genesee, Fulton, Montgomery, Greene, Tioga, Franklin, Chenango, Cortland, Allegany, Delaware, Wyoming, Orleans, Essex, Seneca, Schoharie, Lewis, Yates, Schuyler, Hamilton)
base_url = https://health.data.ny.gov/resource/xdss-u53e.json?County=
urls = Array(https://health.data.ny.gov/resource/xdss-u53e.json?County=Kings, https://health.d...


Array(https://health.data.ny.gov/resource/xdss-u53e.json?County=Kings, https://health.d...

In [3]:
val data_rdd_2 = spark.read.json(sc.parallelize(results,32))
                    .rdd
                    .collect()

data_rdd_2 = Array([Kings,926931,20350479,COUNTY,201,2023-08-30T00:00:00.000,13.80%,1491], [Kings,926730,20348988,COUNTY,165,2023-08-29T00:00:00.000,10.60%,1472], [Kings,926565,20347516,COUNTY,131,2023-08-28T00:00:00.000,13.15%,1032], [Kings,926434,20346484,COUNTY,126,2023-08-27T00:00:00.000,14.02%,910], [Kings,926308,20345574,COUNTY,132,2023-08-26T00:00:00.000,10.71%,1270], [Kings,926176,20344304,COUNTY,168,2023-08-25T00:00:00.000,12.02%,1413], [Kings,926008,20342891,COUNTY,190,2023-08-24T00:00:00.000,14.72%,1383], [Kings,925818,20341508,COUNTY,179,2023-08-23T00:00:00.000,12.78%,1399], [Kings,925639,20340109,COUNTY,198,2023-08-22T00:00:00.000,11.29%,1672], [Kings,925441,20338437,COUNTY,122,2023-08-21T00:00:00.000,13.96%,929], [Kings,925319,20337508,COUN...


Array([Kings,926931,20350479,COUNTY,201,2023-08-30T00:00:00.000,13.80%,1491], [Kings,926730,20348988,COUNTY,165,2023-08-29T00:00:00.000,10.60%,1472], [Kings,926565,20347516,COUNTY,131,2023-08-28T00:00:00.000,13.15%,1032], [Kings,926434,20346484,COUNTY,126,2023-08-27T00:00:00.000,14.02%,910], [Kings,926308,20345574,COUNTY,132,2023-08-26T00:00:00.000,10.71%,1270], [Kings,926176,20344304,COUNTY,168,2023-08-25T00:00:00.000,12.02%,1413], [Kings,926008,20342891,COUNTY,190,2023-08-24T00:00:00.000,14.72%,1383], [Kings,925818,20341508,COUNTY,179,2023-08-23T00:00:00.000,12.78%,1399], [Kings,925639,20340109,COUNTY,198,2023-08-22T00:00:00.000,11.29%,1672], [Kings,925441,20338437,COUNTY,122,2023-08-21T00:00:00.000,13.96%,929], [Kings,925319,20337508,COUN...

In [4]:
//Check the data_start_date and data_end_date
val data_start_date = data_rdd.map(_._1).min()
val data_end_date = data_rdd.map(_._1).max()

data_start_date = 2020-03-05
data_end_date = 2023-08-30


2023-08-30

In [5]:
data_rdd.collect()

Array((2023-08-30,Kings,201,1491), (2023-08-29,Kings,165,1472), (2023-08-28,Kings,131,1032), (2023-08-27,Kings,126,910), (2023-08-26,Kings,132,1270), (2023-08-25,Kings,168,1413), (2023-08-24,Kings,190,1383), (2023-08-23,Kings,179,1399), (2023-08-22,Kings,198,1672), (2023-08-21,Kings,122,929), (2023-08-20,Kings,95,769), (2023-08-19,Kings,132,1152), (2023-08-18,Kings,144,1242), (2023-08-17,Kings,159,1337), (2023-08-16,Kings,161,1384), (2023-08-15,Kings,148,1348), (2023-08-14,Kings,93,820), (2023-08-13,Kings,110,893), (2023-08-12,Kings,118,1104), (2023-08-11,Kings,120,1248), (2023-08-10,Kings,149,1192), (2023-08-09,Kings,143,1155), (2023-08-08,Kings,134,1191), (2023-08-07,Kings,101,805), (2023-08-06,Kings,81,702), (2023-08-05,Kings,102,1149), (2023...

In [7]:
// Data Acquisition (Fetching JSON from API)
val data_rdd_new = data_rdd.filter{ case (date, county, cases, tests) =>
  !date.startsWith("2020-03")
}


data_rdd_new = MapPartitionsRDD[29] at filter at <console>:25


MapPartitionsRDD[29] at filter at <console>:25

In [9]:
data_rdd_new.collect()

Array((2023-08-30,Kings,201,1491), (2023-08-29,Kings,165,1472), (2023-08-28,Kings,131,1032), (2023-08-27,Kings,126,910), (2023-08-26,Kings,132,1270), (2023-08-25,Kings,168,1413), (2023-08-24,Kings,190,1383), (2023-08-23,Kings,179,1399), (2023-08-22,Kings,198,1672), (2023-08-21,Kings,122,929), (2023-08-20,Kings,95,769), (2023-08-19,Kings,132,1152), (2023-08-18,Kings,144,1242), (2023-08-17,Kings,159,1337), (2023-08-16,Kings,161,1384), (2023-08-15,Kings,148,1348), (2023-08-14,Kings,93,820), (2023-08-13,Kings,110,893), (2023-08-12,Kings,118,1104), (2023-08-11,Kings,120,1248), (2023-08-10,Kings,149,1192), (2023-08-09,Kings,143,1155), (2023-08-08,Kings,134,1191), (2023-08-07,Kings,101,805), (2023-08-06,Kings,81,702), (2023-08-05,Kings,102,1149), (2023...

**Question 2**
* Using ***reduce*** calculate the total number of cases and total number of tests in New York State during the time period of the data

In [10]:
// Total COVID Cases and Tests (Statewide Summary)
val total_cases_tests = data_rdd_new
  // .map{ case (_, _, cases, tests) => (cases, tests) }
  .map{ t=>(t._3, t._4) }
  .reduce{ (a, b) => (a._1 + b._1, a._2 + b._2) }
                                

total_cases_tests = (6038192,110852985)


(6038192,110852985)

In [7]:
// County-Level Aggregation
val cases_tests_by_county = data_rdd_new
  .map{ case (_, county, cases, tests) => (county, (cases, tests)) } // country is key
  .reduceByKey{ (a, b) => (a._1 + b._1, a._2 + b._2) }              
  .collect()                                                         


cases_tests_by_county = Array((Oneida,(64266,1288046)), (Tompkins,(26095,2337566)), (Chemung,(23831,416716)), (Schenectady,(41689,720900)), (Cattaraugus,(18304,255334)), (Greene,(10503,154563)), (Wyoming,(9455,131344)), (Columbia,(13183,204214)), (Chenango,(11450,204918)), (Ulster,(40625,750008)), (Clinton,(22901,368708)), (Wayne,(21303,335956)), (Herkimer,(16903,280875)), (Nassau,(495119,7483941)), (Seneca,(7573,124265)), (Lewis,(7045,91920)), (Broome,(54562,1012029)), (Erie,(245851,3555714)), (Livingston,(13857,241401)), (Bronx,(453220,8198879)), (Allegany,(9831,215722)), (Queens,(784218,13599435)), (Jefferson,(27344,382040)), (Orleans,(10112,140019)), (Putnam,(29047,438613)), (Rensselaer,(40603,754160)), (Ontario,(25466,418190)), (Suffolk,(512097,7595735)...


Array((Oneida,(64266,1288046)), (Tompkins,(26095,2337566)), (Chemung,(23831,416716)), (Schenectady,(41689,720900)), (Cattaraugus,(18304,255334)), (Greene,(10503,154563)), (Wyoming,(9455,131344)), (Columbia,(13183,204214)), (Chenango,(11450,204918)), (Ulster,(40625,750008)), (Clinton,(22901,368708)), (Wayne,(21303,335956)), (Herkimer,(16903,280875)), (Nassau,(495119,7483941)), (Seneca,(7573,124265)), (Lewis,(7045,91920)), (Broome,(54562,1012029)), (Erie,(245851,3555714)), (Livingston,(13857,241401)), (Bronx,(453220,8198879)), (Allegany,(9831,215722)), (Queens,(784218,13599435)), (Jefferson,(27344,382040)), (Orleans,(10112,140019)), (Putnam,(29047,438613)), (Rensselaer,(40603,754160)), (Ontario,(25466,418190)), (Suffolk,(512097,7595735)...

In [12]:
// Monthly Positivity Rate (Statewide)
val positivity_by_month = data_rdd_new
  .map{ case (date, _, cases, tests) =>
    val ym = date.slice(0,7)        // take only "YYYY-MM"
    (ym, (cases, tests))
  }

  .reduceByKey{ (a, b) =>
    (a._1 + b._1, a._2 + b._2) //for the same month
  }

  .map{ case (ym, (cases, tests)) =>
    val positivity = (cases.toDouble / tests) * 100
    (ym, Math.round(positivity * 100).toDouble / 100)
  }
  .sortBy(_._2, ascending = false)
  .collect()


positivity_by_month = Array((2022-01,15.09), (2023-08,14.12), (2021-12,10.54), (2022-07,9.54), (2023-07,8.82), (2022-12,8.07), (2022-05,7.87), (2023-01,7.49), (2022-08,7.08), (2022-09,6.94), (2022-11,6.93), (2022-10,6.82), (2021-01,6.27), (2022-06,5.88), (2020-12,5.82), (2022-04,5.45), (2023-06,4.89), (2023-02,4.63), (2021-02,3.66), (2021-11,3.55), (2021-03,3.35), (2021-08,3.15), (2022-02,2.93), (2021-09,2.9), (2021-04,2.69), (2023-05,2.67), (2023-03,2.62), (2021-10,2.31), (2022-03,1.95), (2023-04,1.92), (2021-07,1.56), (2021-05,1.09), (2021-06,0.42))


Array((2022-01,15.09), (2023-08,14.12), (2021-12,10.54), (2022-07,9.54), (2023-07,8.82), (2022-12,8.07), (2022-05,7.87), (2023-01,7.49), (2022-08,7.08), (2022-09,6.94), (2022-11,6.93), (2022-10,6.82), (2021-01,6.27), (2022-06,5.88), (2020-12,5.82), (2022-04,5.45), (2023-06,4.89), (2023-02,4.63), (2021-02,3.66), (2021-11,3.55), (2021-03,3.35), (2021-08,3.15), (2022-02,2.93), (2021-09,2.9), (2021-04,2.69), (2023-05,2.67), (2023-03,2.62), (2021-10,2.31), (2022-03,1.95), (2023-04,1.92), (2021-07,1.56), (2021-05,1.09), (2021-06,0.42))

In [14]:
//Daily Positivity per County (Safe Division & Option)

// It handles division by zero safely
def divide(numerator: Int, denominator: Int): Option[Double] = {
  if (denominator == 0) None
  else Some(numerator.toDouble / denominator)
}



divide: (numerator: Int, denominator: Int)Option[Double]


In [23]:

val positivity = data_rdd_new.map{
  case (date, county, cases, tests) =>

    val ratio = divide(cases, tests)

    // convert Option[Double] to a number (0.0 if None)
    val positivity_value = ratio match {
      case Some(r) => Math.round(r * 10000).toDouble / 100 //keep 2 decimal if Math.round(r * 100).toDouble would be 13.0
      case None    => 0.0
    }

    // return (date, county, positivity)
    (date, county, positivity_value)
}
.collect()



positivity = Array((2023-08-30,Kings,13.48), (2023-08-29,Kings,11.21), (2023-08-28,Kings,12.69), (2023-08-27,Kings,13.85), (2023-08-26,Kings,10.39), (2023-08-25,Kings,11.89), (2023-08-24,Kings,13.74), (2023-08-23,Kings,12.79), (2023-08-22,Kings,11.84), (2023-08-21,Kings,13.13), (2023-08-20,Kings,12.35), (2023-08-19,Kings,11.46), (2023-08-18,Kings,11.59), (2023-08-17,Kings,11.89), (2023-08-16,Kings,11.63), (2023-08-15,Kings,10.98), (2023-08-14,Kings,11.34), (2023-08-13,Kings,12.32), (2023-08-12,Kings,10.69), (2023-08-11,Kings,9.62), (2023-08-10,Kings,12.5), (2023-08-09,Kings,12.38), (2023-08-08,Kings,11.25), (2023-08-07,Kings,12.55), (2023-08-06,Kings,11.54), (2023-08-05,Kings,8.88), (2023-08-04,Kings,9.6), (2023-08-03,Kings,14.76), (2023-08-02,Kings,10.3...


Array((2023-08-30,Kings,13.48), (2023-08-29,Kings,11.21), (2023-08-28,Kings,12.69), (2023-08-27,Kings,13.85), (2023-08-26,Kings,10.39), (2023-08-25,Kings,11.89), (2023-08-24,Kings,13.74), (2023-08-23,Kings,12.79), (2023-08-22,Kings,11.84), (2023-08-21,Kings,13.13), (2023-08-20,Kings,12.35), (2023-08-19,Kings,11.46), (2023-08-18,Kings,11.59), (2023-08-17,Kings,11.89), (2023-08-16,Kings,11.63), (2023-08-15,Kings,10.98), (2023-08-14,Kings,11.34), (2023-08-13,Kings,12.32), (2023-08-12,Kings,10.69), (2023-08-11,Kings,9.62), (2023-08-10,Kings,12.5), (2023-08-09,Kings,12.38), (2023-08-08,Kings,11.25), (2023-08-07,Kings,12.55), (2023-08-06,Kings,11.54), (2023-08-05,Kings,8.88), (2023-08-04,Kings,9.6), (2023-08-03,Kings,14.76), (2023-08-02,Kings,10.3...

In [27]:

val positivity_1 = data_rdd_new.map{
  case (date, county, cases, tests) =>

    val ratio = divide(cases, tests)

    // convert Option[Double] to a number (0.0 if None)
    val positivity_value = ratio match {
      case Some(r) => Math.round(r * 10000).toDouble / 100 //keep 2 decimal if Math.round(r * 100).toDouble would be 13.0
      case None    => 0.0
    }

    // return (date, county, positivity)
    (date, county, positivity_value)
}




positivity_1 = MapPartitionsRDD[39] at map at <console>:25


MapPartitionsRDD[39] at map at <console>:25

In [28]:
//Highest Positive Day in Dataset
val highest_positive = positivity_1.takeOrdered(1)(Ordering[Double].reverse.on(_._3))


highest_positive = Array((2023-07-02,Tompkins,100.0))


Array((2023-07-02,Tompkins,100.0))

In [29]:
//Monthly Positivity Rate by County
val county_month_rdd = data_rdd_new.map{
  case (date, county, cases, tests) =>
    val month = date.slice(0,7)      // get "YYYY-MM"
    ((county, month), (cases, tests))
}

val summed_rdd = county_month_rdd.reduceByKey{
  case ((cases1, tests1), (cases2, tests2)) =>
    (cases1 + cases2, tests1 + tests2)
}

val positivity_rdd = summed_rdd.map{
  case ((county, month), (cases, tests)) =>
    val positivity = if (tests == 0) 0.0 else (cases.toDouble / tests)
    (county, month, Math.round(positivity * 100).toDouble / 100) 
}

val sorted_rdd = positivity_rdd.sortBy(r => (r._1, r._2))

sorted_rdd.collect().foreach{
  case (county, month, positivity) =>
    println(s"$county $month $positivity")
}


Albany 2020-12 0.08
Albany 2021-01 0.08
Albany 2021-02 0.02
Albany 2021-03 0.02
Albany 2021-04 0.02
Albany 2021-05 0.01
Albany 2021-06 0.0
Albany 2021-07 0.02
Albany 2021-08 0.05
Albany 2021-09 0.04
Albany 2021-10 0.03
Albany 2021-11 0.05
Albany 2021-12 0.09
Albany 2022-01 0.15
Albany 2022-02 0.05
Albany 2022-03 0.03
Albany 2022-04 0.09
Albany 2022-05 0.12
Albany 2022-06 0.07
Albany 2022-07 0.11
Albany 2022-08 0.1
Albany 2022-09 0.1
Albany 2022-10 0.1
Albany 2022-11 0.06
Albany 2022-12 0.07
Albany 2023-01 0.09
Albany 2023-02 0.07
Albany 2023-03 0.07
Albany 2023-04 0.05
Albany 2023-05 0.05
Albany 2023-06 0.05
Albany 2023-07 0.08
Albany 2023-08 0.13
Allegany 2020-12 0.09
Allegany 2021-01 0.07
Allegany 2021-02 0.01
Allegany 2021-03 0.0
Allegany 2021-04 0.02
Allegany 2021-05 0.02
Allegany 2021-06 0.01
Allegany 2021-07 0.01
Allegany 2021-08 0.03
Allegany 2021-09 0.05
Allegany 2021-10 0.08
Allegany 2021-11 0.11
Allegany 2021-12 0.12
Allegany 2022-01 0.13
Allegany 2022-02 0.08
Allegany 2022-0

county_month_rdd = MapPartitionsRDD[41] at map at <console>:25
summed_rdd = ShuffledRDD[42] at reduceByKey at <console>:31
positivity_rdd = MapPartitionsRDD[43] at map at <console>:36
sorted_rdd = MapPartitionsRDD[48] at sortBy at <console>:42


MapPartitionsRDD[48] at sortBy at <console>:42

# Conclusion

This Spark-based analysis demonstrates the use of RDD transformations,
key-value aggregations, safe numerical computation, and time-based grouping
to understand COVID-19 positivity trends across New York State.

Key findings include:
- Highest statewide positivity month: 2022-01 (15.09%)
- Highest county positivity day: Tompkins on 2023-07-02 (100%)
- Counties show different monthly trends, indicating localized outbreaks.

